### PR TITLE
Added TCP NS flag, reserved flags.

### PR DESCRIPTION
**Pull request by dougives**
Hello friends,

I added support for changing the NS flag, along with the three reserved flags. I named the reserved flags FlagA, FlagB, FlagC, after their position in the field, but RsvdA or something may be more appropriate.

Also, I imagine that this may break some user code that expects AllFlags to be a byte value, if the user code fails to check the field's size.

It's not terribly useful, but I just wanted some support for it in my program.

### CODE REVIEW

File: PacketDotNet/TcpFields.cs [hunk 1]
```diff
@@ -29,14 +29,18 @@ public struct TcpFields
...
         // flag bitmasks
-        public readonly static int TCP_CWR_MASK = 0x0080;
-        public readonly static int TCP_ECN_MASK = 0x0040;
-        public readonly static int TCP_URG_MASK = 0x0020;
-        public readonly static int TCP_ACK_MASK = 0x0010;
-        public readonly static int TCP_PSH_MASK = 0x0008;
-        public readonly static int TCP_RST_MASK = 0x0004;
-        public readonly static int TCP_SYN_MASK = 0x0002;
-        public readonly static int TCP_FIN_MASK = 0x0001;
+        public readonly static int TCP_FLAGC_MASK = 0x0800;
```
Comment (chmorgan) [inline PacketDotNet/TcpFields.cs]: Why reverse the order of the masks here? It would be clearer if you added the new fields in the style of the existing masks, or separated out reversing the order of the masks into a separate commit, and then introduced the new fields. Doing both means I have to expend effort to distinguish between what you added and what you reordered.

File: PacketDotNet/TcpFields.cs [hunk 2]
```diff
@@ -85,7 +89,7 @@ static TcpFields()
             SequenceNumberPosition = DestinationPortPosition + PortLength;
             AckNumberPosition = SequenceNumberPosition + SequenceNumberLength;
             DataOffsetPosition = AckNumberPosition + AckNumberLength;
-            FlagsPosition = DataOffsetPosition + DataOffsetLength;
+            FlagsPosition = DataOffsetPosition;// + DataOffsetLength;
```
Comment (chmorgan) [inline PacketDotNet/TcpFields.cs]: I noticed your other commit removed the stray comment. You can rebase this commit and remove the introduction of the comment before you ever add it.

File: Examples/ConstructingPackets/Main.cs [hunk 3]
```diff
@@ -36,6 +36,7 @@ public static void Main(string[] args)
 
             // and print out the packet to see that it looks just like we wanted it to
             Console.WriteLine(ethernetPacket.ToString());
+            Console.ReadKey(true);
```
Comment (chmorgan) [inline Examples/ConstructingPackets/Main.cs]: why this change?

File: PacketDotNet/TcpPacket.cs [hunk 4]
```diff
@@ -294,12 +300,36 @@ virtual public bool CWR
             set { setFlag(value, TcpFields.TCP_CWR_MASK); }
         }
 
+        virtual public bool NS
+        {
+            get { return (AllFlags & TcpFields.TCP_NS_MASK) != 0; }
+            set { setFlag(value, TcpFields.TCP_NS_MASK); }
+        }
+
+        virtual public bool FlagA
```
Comment (chmorgan) [inline PacketDotNet/TcpPacket.cs]: Is there some convention that determined that these should be Flag A, B and C?

Comment (chmorgan) [inline PacketDotNet/TcpPacket.cs]: I'm asking because looking at the docs on wikipedia it isn't clear which of these properties is the highest bit vs. the lowest bit in the reserved flags.

File: Examples/ConstructingPackets/Main.cs (hunk 3 above)
Comment (dougives) [inline Examples/ConstructingPackets/Main.cs]: It was just to keep the window open to view the results of the example. I checked the other examples and they didn't have it, so I will change it to be consistent, and in case anyone decides to scrape it.

File: PacketDotNet/TcpPacket.cs (hunk 4 above)
Comment (dougives) [inline PacketDotNet/TcpPacket.cs]: Not to my knowledge, I looked on wikipedia and the RFCs, and they are just labeled as reserved. I labeled them in order from LSB: A, B, and C. I needed to distinguish these bits, but as far as I know, there is no convention so far.

File: PacketDotNet/TcpFields.cs (hunk 1 above)
Comment (dougives) [inline PacketDotNet/TcpFields.cs]: I'm not sure what you mean. They were originally ordered from MSB to LSB, and I retained the ordering, unless I misunderstand the issue.

Comment (chmorgan) [inline PacketDotNet/TcpFields.cs]: Hmm. Maybe it is related to a line ending change? I just pushed out a bunch of changes to clean up some inconsistent line endings. If you can rebase against master the extra changes might go away. Based on your comment I'd only expect to see 4 new lines added here.

File: Examples/ConstructingPackets/Main.cs (hunk 3 above)
Comment (chmorgan) [inline Examples/ConstructingPackets/Main.cs]: Ahh. I normally run the applications from the console (on linux) but I see what you mean. Windows I think will just close the application window out once the application exits, and it will do so immediately without that line. Its probably best to stay consistent until we can discuss and agree upon the other change. Is it so unusual to run applications from the console on windows?

**Comment by chmorgan** [comment]
Did you see my comments on the overall change? I could take a crack at adding NS flag support and exposing the full set of flags via AllFlags if that's what you are looking for.

**Comment by dougives** [comment]
Yeah, NS support would be the ideal situation. I was generating packets using genetic programming, so the only thing I required was to break out those bits so I could see when it made silly decisions.

I'll review this some more in a bit...

**Comment by chmorgan** [comment]
Let me hack something up here based on your commit. What you have is close
but I'm not sure what to do with those reserved flags. Ok if those aren't
accessible via properties for now?

**Comment by dougives** [comment]
Yeah, I no longer use the code that uses those bits, so it won't be an issue.

File: PacketDotNet/TcpFields.cs (hunk 1 above)
Comment (dougives) [inline PacketDotNet/TcpFields.cs]: I see what you're talking about now ... I'm not sure why the diff is like that but I'll rebase it and see if it changes.

File: Examples/ConstructingPackets/Main.cs (hunk 3 above)
Comment (dougives) [inline Examples/ConstructingPackets/Main.cs]: I believe so. It happens to be the default setting in VS to spawn another console. Given that this example works fine and is five years old, it's probably best to keep it unchanged, in case somebody decided to grep it in a script or something.

**Comment by chmorgan** [comment]
Closing this out as a very similar patch was merged to master.
